crates/bevy_ui/src/layout/mod.rs

```diff
@@ -41,25 +41,20 @@ impl LayoutContext {
     }
 }
 
-#[derive(Debug, Clone, PartialEq, Eq)]
-struct RootNodePair {
-    // The implicit "viewport" node created by Bevy
-    implicit_viewport_node: taffy::node::Node,
-    // The root (parentless) node specified by the user
-    user_root_node: taffy::node::Node,
-}
+type EntityTupleHashMap<V> = hashbrown::HashMap<(Entity, Entity), V, EntityHash>;
 
 #[derive(Resource)]
 pub struct UiSurface {
     entity_to_taffy: EntityHashMap<taffy::node::Node>,
-    camera_entity_to_taffy: EntityHashMap<EntityHashMap<taffy::node::Node>>,
-    camera_roots: EntityHashMap<Vec<RootNodePair>>,
+    /// `HashMap<(Camera Entity, Root (parentless) Entity), implicit Viewport>`
+    camera_root_to_viewport_taffy: EntityTupleHashMap<taffy::node::Node>,
```

Comment:
Few things here:
- `(Entity, Entity)` as the key makes it harder to work with unless you iterate over the whole thing or already have both values. The benefit is that a viewport node can't be inserted without the camera entity and root node entity being in the map as the key.

- We could take this a step further and remove `entity_to_taffy` and restore `RootNodePair`. But that does affect a few bit more LOC. And not to mention, would make lookups even less optimal.

Conclusion; I think these are interesting approaches, but it might be better to create a management struct that maintains it / while preserving the speed as before. Gonna explore more here unless someone has a good suggestion that I'm not seeing!

Reply:
I don't think ` EntityTupleHashMap<taffy::node::Node>` makes sense as you want to be able to iterate over root nodes given just a camera node. I think either of:

1. `EntityHashMap<EntityHashMap<taffy::node::Node>>`
2. `EntityHashMap<Vec<RootNodePair>>`

ought to work fine on their own. They're essentially equivalent with slightly different performance characteristics. 2. is likely faster for iterating and lookups with small numbers of root nodes per camera. 1. is likely faster for lookups with large numbers of root nodes per camera. This is unlikely to be a performance hotspot so we probably ought not to worry about it too much.

A third option of `EntityHashMap<taffy::node::Node>` where the Node stored is the "user root node" and you lookup the "implicit viewport node" using `Taffy.parent(user_root_node)` would probably also work. Looking up a parent in Taffy should be fast as it's just a `Vec` lookup.

Reply:
After a massive 3 day deep dive through the layout system. I've boiled it down to this:

```rust
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct UiNodeMeta {
    pub(super) camera_entity: Option<Entity>,
    pub(super) root_node_pair: RootNodePair,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RootNodePair {
    // The implicit "viewport" node created by Bevy
    pub(super) implicit_viewport_node: taffy::node::Node,
    // The root (parentless) node specified by the user
    pub(super) user_root_node: taffy::node::Node,
}

#[derive(Resource)]
pub struct UiSurface {
    pub(super) entity_to_taffy: EntityHashMap<taffy::node::Node>,
    // I'm not sure I like the name `meta` but can't think of anything else right now. 
    // maybe `relations`?
    pub(super) ui_root_node_meta: EntityHashMap<UiNodeMeta>,
    pub(super) camera_root_nodes: EntityHashMap<EntityHashSet>,
    pub(super) taffy: Taffy,
}
```

My goal in the next day or so is to [break up the work I've been experimenting with](https://github.com/bevyengine/bevy/compare/cb9789bc35ca13d631c5f79e8a7ed4e62a5dd6eb...StrikeForceZero:bevy:dev/cleanup_ui_surface_mappings_experimental) and submit several incremental PR's with intentions to:
- document several areas, including the ones I unknowingly stomped over
- extract `UiSurface` into its own file since `layout/mod` is getting pretty big.
- to clean some stuff up
- implement the revised mapping structure
- add more regression tests for edge cases I found affecting main